### ISSUE TITLE
Reordering of CSC matrix breaks when you go above int32 limits

### ISSUE THREAD

**Issue by tueboesen** [scipy/scipy#12015]
I am trying to solve a linear system as effeciently as possible, and one of the methods I have been trying is by reordering the matrix beforehand with reverse_cuthill_mckee.
This works well for small matrices, but when I move to a large size this breaks. (60000x60000)

The code to reproduce this is really simple:
```
perm = scipy.sparse.csgraph.reverse_cuthill_mckee(H, symmetric_mode=True)
I,J = np.ix_(perm,perm)
H_reordered = H[I,J]
```
Where H is a symmetric 60000x60000 sparse csc matrix with some nonzero elements. In my case it has about 7 million nonzero elements

When I run the above I get:

```
Traceback (most recent call last):
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2020.1\plugins\python-ce\helpers\pydev\pydevd.py", line 1438, in _exec
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2020.1\plugins\python-ce\helpers\pydev\_pydev_imps\_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "D:/Dropbox/ActiveLearning/Pytorch/Active-Learning/pcg.py", line 37, in <module>
    H_ref = H[I,J]
  File "D:\Dropbox\ActiveLearning\Pytorch\Active-Learning\venv\lib\site-packages\scipy\sparse\_index.py", line 75, in __getitem__
    return self._get_arrayXarray(row, col)
  File "D:\Dropbox\ActiveLearning\Pytorch\Active-Learning\venv\lib\site-packages\scipy\sparse\compressed.py", line 664, in _get_arrayXarray
    csr_sample_values(M, N, self.indptr, self.indices, self.data,
ValueError: could not convert integer scalar
```

And the reason as far as I can see is that the reordering involves making a 60000x60000 ndarray, which has about 3.6*10^9 elements and hence is too big int32.

Does anyone know of a workaround for this problem? or a fix?

**Comment by tueboesen** [comment]
I should note that the above method works without problems for smaller matrices

**Comment by rgommers** [comment]
@tueboesen your example isn't self-contained (`H` is undefined), but `reverse_cuthill_mckee` does support `int64`, can you make sure your `H` uses `int64` for `H.indices` and `H.indptr`?

**Comment by tueboesen** [comment]
Here is a self-contained example that shows it:

```
import scipy
import numpy as np
from scipy.sparse import rand

n = 60000
x = np.random.randn(n)

H = scipy.sparse.dia_matrix((x,0),(n,n))

H = H.tocsc()

perm = scipy.sparse.csgraph.reverse_cuthill_mckee(H, symmetric_mode=True)
I, J = np.ix_(perm, perm)
H_reordered = H[I, J]
```

**Comment by ewmoore** [comment]
Don't use `np.ix_`.  Just permute the rows then the columns directly using the results of `reverse_cuthill_mckee`.   Doing it with the output of `np.ix_` means realizing a dense intermediate, even if your indexing arrays are sparse.

```
import scipy
import numpy as np
from scipy.sparse import rand

n = 60000
x = np.random.randn(n)

H = scipy.sparse.dia_matrix((x,0),(n,n))

H = H.tocsc()

perm = scipy.sparse.csgraph.reverse_cuthill_mckee(H, symmetric_mode=True)
H_reordered = H[perm,:][:,perm]
```

Maybe we need to add an example to `reverse_cuthill_mckee` with some text indicating how you should apply the permutation to a sparse matrix?

**Comment by ewmoore** [comment]
This is a duplicate of #11255 and is fixed in master.  

On master: 
```
In [1]: import numpy as np

In [2]: from scipy import sparse

In [3]: %load_ext memory_profiler

In [4]: s = sparse.random(10000, 10000, format='csr')

In [5]: idx = np.arange(10000)

In [6]: %memit s[idx.reshape(-1, 1), idx]
peak memory: 100.10 MiB, increment: 11.83 MiB

In [7]: %memit s[np.ix_(idx, idx)]
peak memory: 100.10 MiB, increment: 0.00 MiB

In [8]: s[idx.reshape(-1, 1), idx] == s[np.ix_(idx, idx)]
/home/ewm/venv/scipy/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3331: SparseEfficiencyWarning: Comparing sparse matrices using == is inefficient, try using != instead.
  exec(code_obj, self.user_global_ns, self.user_ns)
Out[8]:
<10000x10000 sparse matrix of type '<class 'numpy.bool_'>'
        with 100000000 stored elements in Compressed Sparse Row format>

In [9]: np.all(_8.data)
Out[9]: True
```
The fix was #11335.
